--- Day 1: Trebuchet?! ---
Something is wrong with global snow production, and you've been selected to take a look. The Elves have even given you a map; on it, they've used stars to mark the top fifty locations that are likely to be having problems.

You've been doing this long enough to know that to restore snow operations, you need to check all fifty stars by December 25th.

Collect stars by solving puzzles. Two puzzles will be made available on each day in the Advent calendar; the second puzzle is unlocked when you complete the first. Each puzzle grants one star. Good luck!

You try to ask why they can't just use a weather machine ("not powerful enough") and where they're even sending you ("the sky") and why your map looks mostly blank ("you sure ask a lot of questions") and hang on did you just say the sky ("of course, where do you think snow comes from") when you realize that the Elves are already loading you into a trebuchet ("please hold still, we need to strap you in").

As they're making the final adjustments, they discover that their calibration document (your puzzle input) has been amended by a very young Elf who was apparently just excited to show off her art skills. Consequently, the Elves are having trouble reading the values on the document.

The newly-improved calibration document consists of lines of text; each line originally contained a specific calibration value that the Elves now need to recover. On each line, the calibration value can be found by combining the first digit and the last digit (in that order) to form a single two-digit number.

For example:

1abc2
pqr3stu8vwx
a1b2c3d4e5f
treb7uchet
In this example, the calibration values of these four lines are 12, 38, 15, and 77. Adding these together produces 142.

Consider your entire calibration document. What is the sum of all of the calibration values?

In [93]:
def loadContent(filePath):
    with open(filePath, 'r', encoding='utf-8') as file:
        content = file.read()

    return content.split()

In [94]:
def getNumbers(contents):
    result_dict = {}
    for line in contents:
        result_dict[line] = getNumber(line)

    return result_dict

def getNumber(line):
    result = []
    for char in line:
        if char.isdigit():
            result.append(char)
    first_no = result[0]
    last_index = len(result) - 1
    if last_index >= 0:
        last_no = result[last_index]
    else:
        last_no = result[0]
    
    return int(first_no + last_no)

In [95]:
contents = loadContent('input.txt')
result_dict = getNumbers(contents)
finall_result = sum(result_dict.values())
finall_result

53386

Your puzzle answer was 53386.

The first half of this puzzle is complete! It provides one gold star: *

--- Part Two ---
Your calculation isn't quite right. It looks like some of the digits are actually spelled out with letters: one, two, three, four, five, six, seven, eight, and nine also count as valid "digits".

Equipped with this new information, you now need to find the real first and last digit on each line. For example:

two1nine
eightwothree
abcone2threexyz
xtwone3four
4nineeightseven2
zoneight234
7pqrstsixteen
In this example, the calibration values are 29, 83, 13, 24, 42, 14, and 76. Adding these together produces 281.

What is the sum of all of the calibration values?

In [121]:
import re

def replace_strings(text):
    text = re.sub(r'on', 'oon', text)
    text = re.sub(r'ei', 'eei', text)
    text = re.sub(r'ht', 'htt', text)

    return text

def getPattern(words_to_numbers):
    return re.compile(r'(' + '|'.join(re.escape(key) for key in words_to_numbers.keys()) + r'|[0-9])')

def getTranslatedNumbers(contents):
    words_to_numbers = {
        "one": "1", "two": "2", "three": "3", "four": "4",
        "five": "5", "six": "6", "seven": "7", "eight": "8",
        "nine": "9", 
    }
    pattern = getPattern(words_to_numbers)
    
    result_dict = {}
    for line in contents:
        corrected_line = replace_strings(line)
        matches = [match.group() for match in re.finditer(pattern, corrected_line)]

        translated_matches = [words_to_numbers.get(match, match) for match in matches]
        translated_text = "".join(translated_matches)

        number = getNumber(translated_text)
        result_dict[line] = (corrected_line, matches, translated_text, number)

    return result_dict

In [137]:
#contents = loadContent('input.txt')
contents2 = ["xtwone3four"]

result_dict = getTranslatedNumbers(contents)
print(result_dict)

numbers = [value[len(value) - 1] for value in result_dict.values()]
print(numbers)

finall_result = sum(numbers)
finall_result

{'ckmb52fldxkseven3fkjgcbzmnr7': ('ckmb52fldxkseven3fkjgcbzmnr7', ['5', '2', 'seven', '3', '7'], '52737', 57), 'gckhqpb6twoqnjxqplthree2fourkspnsnzxlz1': ('gckhqpb6twoqnjxqplthree2fourkspnsnzxlz1', ['6', 'two', 'three', '2', 'four', '1'], '623241', 61), '2onetwocrgbqm7': ('2oonetwocrgbqm7', ['2', 'one', 'two', '7'], '2127', 27), 'frkh2nineqmqxrvdsevenfive': ('frkh2nineqmqxrvdsevenfive', ['2', 'nine', 'seven', 'five'], '2975', 25), 'four9two': ('four9two', ['four', '9', 'two'], '492', 42), 'six7sixqrdfive3twonehsk': ('six7sixqrdfive3twoonehsk', ['six', '7', 'six', 'five', '3', 'two', 'one'], '6765321', 61), 'xkvsone2': ('xkvsoone2', ['one', '2'], '12', 12), 'one65': ('oone65', ['one', '6', '5'], '165', 15), 'rggxsff1seven': ('rggxsff1seven', ['1', 'seven'], '17', 17), 'djbcgrrtqdshpqqzj43rgcr': ('djbcgrrtqdshpqqzj43rgcr', ['4', '3'], '43', 43), 'br89fivetwoqggnxjfourtl3': ('br89fivetwoqggnxjfourtl3', ['8', '9', 'five', 'two', 'four', '3'], '895243', 83), 'zoneight47five5sixjxd74': ('zoo

53312

Your puzzle answer was 53312.

Both parts of this puzzle are complete! They provide two gold stars: **